### PR TITLE
Private events

### PR DESCRIPTION
**Pull request by Alexoid1**
## In this project, I build a site similar to a private Eventbrite with authentication and association between components.

- User authentication without use the divise gem.
- Seesions.
- Association many through many
- Set bootstrap in a rails app.
- Use helpers and scopes.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    @users = User.all
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def show
+    @user = User.find(params[:id])
+    @upcoming
```
Comment (maelfosso) [inline app/controllers/users_controller.rb]: That variable seems not to be used in the view, kindly remove it

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+<div class='border col-8 mx-auto mt-3'>
+    <h5 class='text-center'>Upcoming events</h5>
+    <hr class="my-4">
+    <div class='row d-flex justify-content-around'>
+        <% @user.attended_events.each do |event| %>
+            <%= uncoming(event)  %>
+            
+        <% end %>
+    </div>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: It's not the good way to implement it.
Kindly use `@user.attended_events.upcoming`

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    
+</div>
+<div class='border col-8 mx-auto mt-3'>
+    <h5 class='text-center'>Previous events</h5>
+    <hr class="my-4">
+    <div class='row d-flex justify-content-around'>
+        <% @user.attended_events.each do |event| %>
+            <%= previous(event)  %>
+            
+        <% end %>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: It's not a good way to implement it.
Kindly use `@user.attended_events.previous`

File: app/helpers/users_helper.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  def previous(event)
+    return unless Date.parse(event.time) <= Date.today
+
+    content_tag :div, class: 'border col-3 m-3' do
+      content_tag(:h5, event.title, class: 'text-center') +
+        content_tag(:p, event.description) +
+        content_tag(:p, 'Location: ' + event.location) +
+        content_tag(:p, 'Date: ' + event.time)
+    end
+  end
```
Comment (maelfosso) [inline app/helpers/users_helper.rb]: It's not needed.
Kindly read the other comments to know why

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (maelfosso) [reply app/controllers/users_controller.rb]: Kindly create two view variables for past and upcoming attended events
```ruby
@upcoming = @user.attended_events.upcoming
@previous = @user.attended_events.previous
```

File: app/helpers/events_helper.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+module EventsHelper
+  def image_event(event)
+    image_tag event.image.url :medium if event.image.exists?
+  end
+
+  def created_by(event)
+    @users.find_by_id(event.user_id).name
+  end
```
Comment (maelfosso) [inline app/helpers/events_helper.rb]: That helper is not needed
You can directly access to the event author with `event.user.name`

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+class UsersController < ApplicationController
```
Comment (maelfosso) [inline app/controllers/users_controller.rb:1]: Protect that controller. 
Only connected user must be able to have access to its actions
Use `before_action` to do that

File: app/controllers/events_controller.rb [hunk 7]
```diff
@@ -0,0 +1,86 @@
+class EventsController < ApplicationController
```
Comment (maelfosso) [inline app/controllers/events_controller.rb:1]: Protect that controller. 
Only connected user must be able to have access to its actions
Use `before_action` to do that

Comment (maelfosso) [reply app/controllers/events_controller.rb:1]: It's possible to access to the event creation form without being connected even if it's generated an error
```rb

20 21 22 23 24 25 | # GET /events/new   def new     @event = current_user.events.build   end    # GET /events/1/edit
-- | --



```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle 

Hello
Thank you for the job done so far.
Everything is working like a charm. However, there are some issues related to the code that needs to be highlighted and solved.

Kindly read the comments left below to know more about the errors to solved
Once done submit a new code review request

Thanks

Mael FOSSO

File: Gemfile [hunk 8]
```diff
@@ -16,7 +16,14 @@ gem 'webpacker', '~> 4.0'
 # Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks
 gem 'turbolinks', '~> 5'
 # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
+gem 'hirb', '~> 0.7.3'
 gem 'jbuilder', '~> 2.7'
+
+gem 'bcrypt-ruby', '~> 3.1', '>= 3.1.5'
```
Comment (christianotieno) [inline Gemfile]: ## Name change
- [x] The name for the gem has changed. Would you mind changing it to the suggestion below for efficiency in the dependencies installation process, please?
```suggestion
gem 'bcrypt', '~> 3.1', '>= 3.1.16'
```

File: README.md [hunk 9]
```diff
@@ -1,24 +1,71 @@
...
-* Configuration
+## Getting Started
+- Go to (https://github.com/Alexoid1/Private-events) and have a look around.
+- Open your Terminal.
+- Install Imagemagick (brew install imagemagick)
+- Clone repo
+- Go to the project's root
+- Type (bundle install)
+- Start server (rails server).
+- In your browser open: http://localhost:3000/
```
Comment (christianotieno) [inline README.md:23]: ## More elaborate instructions
- [ ] Add instructions on how to run your tests.
- [ ] Add a live link to where the site is live (hosted) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_readme.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/)). Thank you.

**Review by christianotieno — CHANGES_REQUESTED** [review]
# Changes Required ♻️ 

Hello @Alexoid1 👋 

Good job working on the suggested changes, kindly take note of the following before we finish things up here. 👇 

#### Christian Otieno (TSE)

**Review by maelfosso — APPROVED** [review]
# Congratulations :clap:

Thank you for applying the suggested changes
You can merge it now :+1: 

Good continuation :handshake: 

Mael FOSSO

**Pull request merged**